### DISCUSSION TITLE
How to achieve a event system?

### DISCUSSION THREAD

**Discussion by Imitater967** [sebas77/Svelto.ECS#111] [Q&A] [answered]
How to achieve a event system if we were unable to create or delete component at runtime?

In unity ECS, the event system is like adding a event component which containing required data, and there is a system to process this event.

for example,

block: entity
- Position: Component
- BlockType: Component
then a block interact system was called, and it would become

bock: entity
- Position: Component
- BlockInteractEvent: Component.
- BlockType: Component
- 
a system called BlockInteractEventSystem would process any entities with comp BlockType and comp BlockInteractEvent
and remove the comp BlockInteractEvent after finish

**Comment by sebas77** [comment, marked as answer]
Hello,

in DOTS ECS, when you add a new component to an entity, you are effectively moving the entity to a different subset. This is indeed the way to handle change of behaviours. If a new behaviour must be applied to an entity, it should be found into another subset.

In Svelto you can create subsets with groups and filters. GroupCompounds help handing multiple states at once. A group compound handles up to 4 states, where the first 1 is usually the type of entity itself. So usually you can handle 3 different states at once. 
While you can iterate entities found in specific groups, it's also possible to rely on the **IReactOnSwapEx callbacks** to react on specific change of state once.

When this becomes to restrictive or cumbersome, filters can be used. the new _filters API_ allows to iterate entities found in specific filters. Each entity can be in any filter you need. the filter ID can identify a state.

Rarely is necessary to communicate that something happens and this is awkward to do both with groups and filters. In this case you could potentially create new entities when an event happens and rely on **IReactOnAddEx callbacks**
